Sources/StreamChat/Workers/EventNotificationCenter.swift

```diff
@@ -33,61 +29,27 @@ class EventNotificationCenter: NotificationCenter {
         middlewares.append(middleware)
     }
 
-    func process(_ event: Event) {
-        var shouldScheduleProcessing = false
-        _pendingEvents {
-            shouldScheduleProcessing = $0.isEmpty
-            $0.append(event)
-        }
+    func process(_ events: [Event], post: Bool = true, completion: (() -> Void)? = nil) {
+        var eventsToPost = [Event]()
         
-        if shouldScheduleProcessing {
-            scheduleProcessing()
-        }
-    }
-    
-    /// Custom overload for `HealthCheckEvent`
-    ///
-    /// When we receive this event, we first create the `CurrentUserDTO`
-    /// and then inform the `connectionIdWaiters` of `ChatClient`
-    /// that we're connected. Using the normal `process` func, by the time
-    /// we inform the waiters, DB save is not finished so `CurrentUserDTO`
-    /// does not exist in DB yet. Using this overload makes sure that
-    /// we process the event fully before calling completion closures.
-    func process(_ healthCheckEvent: HealthCheckEvent, completion: @escaping ((ConnectionId) -> Void)) {
-        database.write { session in
-            // We don't want to publish `HealthCheckEvent`, so we discard the output
-            _ = self.middlewares.process(event: healthCheckEvent, session: session)
-        } completion: { _ in
-            completion(healthCheckEvent.connectionId)
-        }
-    }
-    
-    /// Starts the timer and schedules the processing of events
-    private func scheduleProcessing() {
-        DispatchQueue.main.asyncAfter(deadline: .now() + eventBatchPeriod) { [weak self] in
-            guard let self = self else { return }
-
-            var eventsToPublish: [Event] = []
-            self.database.write({ session in
-                var eventsToProcess: [Event] = []
-                self._pendingEvents {
-                    eventsToProcess = $0
-                    $0.removeAll()
+        database.write({ session in
+            eventsToPost = events.compactMap {
+                self.middlewares.process(event: $0, session: session)
+            }
+        }, completion: { _ in
+            self.eventPostingQueue.sync {
+                if post {
```

Comment:
Why don't we move this `if` check above, to reduce unnecessary `queue.sync` calls?

Reply:
Good stuff, fixed that 👍